prql-compiler/src/semantic/resolver.rs

```diff
@@ -53,169 +73,359 @@ impl AstFold for Resolver {
+                node.declared_at = Some(*id);
+                node
+            })
+            .collect();
+
+        Ok(())
+    }
+
+    fn validate_function_call(
+        &self,
+        declared_at: Option<usize>,
+        mut func_call: FuncCall,
+        is_curry: bool,
+    ) -> Result<(FuncCall, FuncDef), Error> {
+        if declared_at.is_none() {
+            return Err(Error::new(Reason::NotFound {
+                name: func_call.name,
+                namespace: "function".to_string(),
+            }));
+        }
+
+        let func_dec = declared_at.unwrap();
+        let func_dec = &self.context.declarations[func_dec].0;
+        let func_def = func_dec.as_function().unwrap().clone();
+
+        // extract needed named args from positionals
+        let named_params: HashSet<_> = (func_def.named_params)
+            .iter()
+            .map(|param| &param.item.as_named_expr().unwrap().name)
+            .collect();
+        let (named, positional) = func_call.args.into_iter().partition(|arg| {
+            // TODO: replace with drain_filter when it hits stable
+            if let Item::NamedExpr(ne) = &arg.item {
+                named_params.contains(&ne.name)
+            } else {
+                false
+            }
+        });
+        func_call.args = positional;
+        func_call.named_args = named
+            .into_iter()
+            .map(|arg| arg.item.into_named_expr().unwrap())
+            .map(|ne| (ne.name, ne.expr))
+            .collect();
+
+        // validate number of parameters
+        let expected_len = func_def.positional_params.len() - is_curry as usize;
+        let passed_len = func_call.args.len();
+        if expected_len != passed_len {
+            let mut err = Error::new(Reason::Expected {
+                who: Some(func_call.name.clone()),
+                expected: format!("{} arguments", expected_len),
+                found: format!("{}", passed_len),
+            });
+
+            if passed_len > expected_len && passed_len >= 2 {
+                err = err.with_help(format!(
+                    "If you are calling a function, you may want to add braces: `{} [{:?} {:?}]`",
+                    func_call.name, func_call.args[0], func_call.args[1]
+                ));
+            }
+
+            return Err(err);
+        }
+
+        Ok((func_call, func_def))
+    }
+}
+
+fn extract_group_by(nodes: &[Node]) -> Vec<usize> {
+    nodes.iter().map(|n| n.declared_at.unwrap()).collect()
+}
+
+fn extract_sorts(sort: Vec<ColumnSort>) -> Result<Vec<ColumnSort<usize>>> {
+    sort.into_iter()
+        .map(|s| {
+            Ok(ColumnSort {
+                column: (s.column.declared_at)
+                    .ok_or_else(|| anyhow!("Unresolved ident in sort?"))?,
+                direction: s.direction,
+            })
+        })
+        .try_collect()
+}
+
+/// Loads `internal.prql` which contains type definitions of transforms
+pub fn init_context() -> Context {
```

Comment:
WDYT about making this the impl for `Default`? 

---

Zooming out: one thing I worry about is that as the code gets more powerful, it becomes more coupled — it's harder to trace through how a simple query gets parsed, resolved & translated.

I'll think more about this. To some extent, the complication is an inevitable cost of progress, but sometimes there are things we can do so that the simple case is more obvious — possible making default impls is one (I even went through a phase a month ago of reorganizing files so the important functions were at the top of files, as an example of another).

Reply:
I had that as a default trait, but default is usually used for a few assignes to fields and that's it.

This function actually parses a prql file with type definition for transforms, so it feelt too big for a drfault trait. Note that you dont have to call it yourself, it will get called if you pass `contex = None` to resolve function.

Reply:
I agree that modules should be decupled and should have one major function that you can call, and thats it. Thats the objective, that we currently slightly diviate from.